lib/_tls_wrap.js

```diff
@@ -37,211 +37,228 @@ const tls_wrap = process.binding('tls_wrap');
 const TCP = process.binding('tcp_wrap').TCP;
 const Pipe = process.binding('pipe_wrap').Pipe;
 const errors = require('internal/errors');
+const kConnectOptions = Symbol('connect-options');
 const kDisableRenegotiation = Symbol('disable-renegotiation');
+const kErrorEmitted = Symbol('error-emitted');
+const kHandshakeTimeout = Symbol('handshake-timeout');
+const kRes = Symbol('res');
+const kSNICallback = Symbol('snicallback');
+
+const noop = () => {};
 
 function onhandshakestart() {
   debug('onhandshakestart');
 
-  var self = this;
-  var ssl = self._handle;
-  var now = Timer.now();
+  const owner = this.owner;
+  const now = Timer.now();
 
-  assert(now >= ssl.lastHandshakeTime);
+  assert(now >= this.lastHandshakeTime);
 
-  if ((now - ssl.lastHandshakeTime) >= tls.CLIENT_RENEG_WINDOW * 1000) {
-    ssl.handshakes = 0;
+  if ((now - this.lastHandshakeTime) >= tls.CLIENT_RENEG_WINDOW * 1000) {
+    this.handshakes = 0;
   }
 
-  var first = (ssl.lastHandshakeTime === 0);
-  ssl.lastHandshakeTime = now;
+  const first = (this.lastHandshakeTime === 0);
+  this.lastHandshakeTime = now;
   if (first) return;
 
-  if (++ssl.handshakes > tls.CLIENT_RENEG_LIMIT) {
+  if (++this.handshakes > tls.CLIENT_RENEG_LIMIT) {
     // Defer the error event to the next tick. We're being called from OpenSSL's
     // state machine and OpenSSL is not re-entrant. We cannot allow the user's
     // callback to destroy the connection right now, it would crash and burn.
-    setImmediate(function() {
-      var err = new errors.Error('ERR_TLS_SESSION_ATTACK');
-      self._emitTLSError(err);
-    });
+    setImmediate(emitSessionAttackError, owner);
   }
 
-  if (this[kDisableRenegotiation] && ssl.handshakes > 0) {
+  if (owner[kDisableRenegotiation] && this.handshakes > 0) {
     const err = new Error('TLS session renegotiation disabled for this socket');
-    self._emitTLSError(err);
+    owner._emitTLSError(err);
   }
 }
 
+function emitSessionAttackError(socket) {
+  socket._emitTLSError(new errors.Error('ERR_TLS_SESSION_ATTACK'));
+}
 
 function onhandshakedone() {
-  // for future use
   debug('onhandshakedone');
-  this._finishInit();
+
+  const owner = this.owner;
+
+  // `newSession` callback wasn't called yet
+  if (owner._newSessionPending) {
+    owner._securePending = true;
+    return;
+  }
```

Comment:
It's moved from elsewhere. `_finalInit` has two paths: `onhandshakedone` (server) or being called directly. But `_securePending` can only happen on the server because it's called from `onnewsession` which only runs on the server.

```js
    owner._newSessionPending = false;
    if (owner._securePending)
      owner._finishInit();
    owner._securePending = false;
```